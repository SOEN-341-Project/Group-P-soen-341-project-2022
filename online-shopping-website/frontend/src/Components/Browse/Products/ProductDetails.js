import propTypes from 'prop-types';
import { cookieAge } from '../CookieAge';
import Grid from '@mui/material/Grid';
import Card from '@mui/material/Card';
import * as React from 'react';
import { useState } from 'react';
import Button from '@mui/material/Button';
import AddShoppingCartIcon from '@mui/icons-material/AddShoppingCart';
import Stack from '@mui/material/Stack';
import AddIcon from '@mui/icons-material/Add';
import RemoveIcon from '@mui/icons-material/Remove';
import { Link, useParams } from 'react-router-dom';
import axios from 'axios';
import { useCookies } from 'react-cookie';

const ProductButtons = (props) => {

    //Use cookie takes the cookie name as argument and returns the cartCookie object and the setCookie method
    const [cookies, setCookie] = useCookies(["cart", "user"]);

    const [quantity, setQuantity] = useState(1);

    //Forces re-render on call
    const [state, setState] = useState(0);
    const forceUpdate = () => setState(state + 1);

    const productInCart = cookies.cart ? cookies.cart.find(product => props.product.id === product.id) : null;

    const incrementItem = () => {
        //Increment quantity, ensuring that quantity does not exceed maximum 10 items per product in the cart

        if (((productInCart ? productInCart.quantity : 0) + quantity) <= props.product.totalQuantity) {
            setQuantity(quantity + 1);
        }
        forceUpdate();
    }
    const decreaseItem = () => {
        //Decrement quantity, ensuring that quantity has at least 1 item per product in the cart
        if (quantity !== 1) {
            setQuantity(quantity - 1);
        }
        forceUpdate();
    }

    // Modify item's quantity in the cart cookie
    const modifyItemQuantity = (itemId, quantity) => {
        const foundProduct = cookies.cart.find(product => itemId === product.id);
        const newQuantity = foundProduct.quantity + quantity;

        setCookie("cart", cookies.cart.map(product => {
            if (itemId === product.id) {
                return { ...product, quantity: newQuantity };
            }
            return product;
        }), { maxAge: cookieAge });
    }

    const addToCart = () => {
        let item = props.product;

        // Cart is empty, set it to an array containing one product 
        if (!cookies.cart) {
            const newCartItem = {
                id: item.id,
                name: item.name,
                picture: item.picture,
                description: item.description,
                sellerName: item.seller.sellerName,
                sellerId: item.sellerId,
                brandName: item.brand.name,
                brandId: item.brandId,
                price: item.price,
                totalQuantity: item.totalQuantity,
                quantity: quantity
            }

            //setting cookie to the new created item
            setCookie("cart", [newCartItem], { maxAge: cookieAge });
        }

        // Item already in cart
        else if (productInCart) {
            modifyItemQuantity(item.id, quantity);
        }

        // Cart already made, don't have the item
        else {
            const newCartItem = {
                id: item.id,
                name: item.name,
                picture: item.picture,
                description: item.description,
                sellerName: item.seller.sellerName,
                sellerId: item.sellerId,
                brandName: item.brand.name,
                brandId: item.brandId,
                price: item.price,
                totalQuantity: item.totalQuantity,
                quantity: quantity
            }

            //adding item to the cookie array
            cookies.cart.push(
                newCartItem
            );

            setCookie("cart", cookies.cart, { maxAge: cookieAge });
        }
    }

    if (props.product.totalQuantity === 0) {
        return <h3 className='TextPink'>Out of stock</h3>;
    }

    return (
        //Quantity Buttons
        <div className="ProductDetails-QuantityButtonsContainer">
            <h3 className='TextGreen'>Quantity</h3>
            <Stack className="ProductDetails-QuantityButtonsStack" direction="row" spacing={1}>
                <Button className="QuantityButtons-Shared GreenButtonContained" variant="contained"
                    disabled={quantity === 1}
                    onClick={decreaseItem}>
                    <RemoveIcon />
                </Button>
                <input className="inputne" disabled={true} value={quantity} />
                <Button className="QuantityButtons-Shared GreenButtonContained" variant="contained"
                    disabled={((productInCart ? productInCart.quantity : 0) + quantity) >= props.product.totalQuantity}
                    onClick={incrementItem}>
                    <AddIcon />
                </Button>
            </Stack>
            <h5 className="ProductDetails-ProductLimitText">There {props.product.totalQuantity > 1 ? 'are' : 'is'} {props.product.totalQuantity} in stock.</h5>
            {
                // If product is in the user's cart, show quantity of item in cart
                productInCart &&
                <h5 className="ProductDetails-ProductLimitText">You currently have {productInCart.quantity} in your cart.</h5>
            }
            <Button className="ProductDetails-CartButton GreenButtonContained"
                variant="contained"
                endIcon={<AddShoppingCartIcon />}
                // Can't add to cart if logged out, or non-customer, or adding to cart would surpass total quantity
                disabled={
                    !cookies.user
                    || cookies.user.user.role !== 'CUSTOMER'
                    || ((productInCart ? productInCart.quantity : 0) + quantity) > props.product.totalQuantity}
                onClick={addToCart}>
                Add to cart
            </Button>
            {
                // Logged out users may not add to cart
                !cookies.user &&
                <h5 className="ProductDetails-ProductLimitText">You must be logged in to add to cart.</h5>
            }
            {
                // Non-customers may not add to cart
                cookies.user && cookies.user.user.role !== 'CUSTOMER' &&
                <h5 className="ProductDetails-ProductLimitText">Only customers may add to cart.</h5>
            }
        </div>
    );
}
export const ProductDetails = () => {
    //Resetting scrolling to top of the page
    window.scrollTo(0, 0);

    const [selectedProduct, setSelectedProduct] = React.useState(null);
    const [loading, setLoading] = React.useState(true);

    //Getting product name from URL
    const productParams = useParams();

    // Get product by id
    React.useEffect(() => {
        axios.get(process.env.REACT_APP_DB_CONNECTION + "/api/items/find/?id=" + productParams.productId).then((res) => {
            setSelectedProduct(res.data);
            setLoading(false);
        });
    }, [productParams.productId]);

    // Display load screen while getting data
    if (loading) {
        return (
            <Grid container>
                <Grid item xs={12}>
                    <h1 className="TextGreen LoadingSpinnerHeader">Loading product: {productParams.productName}</h1>
                </Grid>
                <Grid item xs={12} id="LoadingSpinner" />
            </Grid>
        );
    }

    return (
        <Grid container className="ProductDetails-Container">
            <Link to="/" className="RoutingLink">
                <Button variant="contained" className="GreenButtonContained">
                    Return to products
                </Button>
            </Link>

            <Grid item container xs={12} className="ProductDetails-ItemContainer">
                <Grid item xs={12} md={6}>
                    <h1>{selectedProduct.name}</h1>
                    <div className="ProductDetails-ImageConatiner">
                        <img className="ProductDetails-Image" src={selectedProduct.picture} alt={selectedProduct.name} />
                    </div>

                    <Grid item container>
                        <Grid item xs={12} md={6}>
                            <h3>Brand</h3>
                            <p>{selectedProduct.brand.name}</p>
                        </Grid>

                        <Grid item xs={12} md={6}>
                            <h3>Seller</h3>
                            <p>{selectedProduct.seller.sellerName}</p>
                        </Grid>
                    </Grid>
                    <h3>Description</h3>
                    <p>{selectedProduct.description}</p>
                </Grid>
                <Grid item md={1} />
                <Grid item xs={12} sm={12} md={5}>
                    <Card className="ProductDetails-SelectionPanel">
                        <h3 className='TextGreen'>Price</h3>
                        <h4 className='TextPink'>{selectedProduct.price} Æ‰</h4>
                        <ProductButtons product={selectedProduct} />
                    </Card>
                </Grid>
            </Grid>
        </Grid>
    );
}

// Check prop types
ProductButtons.propTypes = {
    product: propTypes.shape({
        id: propTypes.number,
        name: propTypes.string,
        picture: propTypes.string,
        description: propTypes.string,
        seller: propTypes.object,
        sellerId: propTypes.number,
        brand: propTypes.object,
        brandId: propTypes.number,
        price: propTypes.number,
        totalQuantity: propTypes.number,
    })
}