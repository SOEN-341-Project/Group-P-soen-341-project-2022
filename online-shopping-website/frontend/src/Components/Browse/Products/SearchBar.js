import React from 'react';
import { IconButton, InputAdornment, TextField } from "@mui/material";
import SearchIcon from '@mui/icons-material/Search';
import CancelIcon from '@mui/icons-material/Cancel';

export const SearchBar = (props) => {
    let filters = props.filters;
    const filterData = props.filterData;

    const [searchValue, setSearchValue] = React.useState('');

    const handleSearchChange = (newValue) => {
        setSearchValue(newValue);
        filters.searchQuery = newValue;
        filterProducts();
    }

    return (
        <div className="SearchBar">
            <TextField
                sx={props.style}
                value={searchValue}
                label="Search Products"
                variant="outlined"
                type="search"
                onChange={(e) => handleSearchChange(e.target.value)}
                fullWidth
                InputProps={{
                    startAdornment: (
                        <InputAdornment position="start">
                            <SearchIcon />
                        </InputAdornment>
                    ),
                    endAdornment: searchValue && (
                        <IconButton
                            aria-label="Clear search"
                            onClick={() => handleSearchChange('')}  
                        >
                            <CancelIcon />
                        </IconButton>
                    )
                }}
            />
        </div>
    );
}