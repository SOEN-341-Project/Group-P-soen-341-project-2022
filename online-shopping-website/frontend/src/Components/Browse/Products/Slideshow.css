.ProductImageShow {
    display: block;
    margin: auto;
    max-height: 200px;
    max-width: 100%;
}

.ProductDetailsShow {
    background-color: white;
    border-radius: 10px;
    margin: 1rem 2.5rem 2.5rem 2.5rem;
    padding: 1rem 2rem;
}
.SlideshowPadding {
    background-color: rgb(62, 134, 62);
    border-radius: 10px;
    color: white;
    margin-bottom: 4rem;
    padding: 1rem;
    text-align: center;
}
