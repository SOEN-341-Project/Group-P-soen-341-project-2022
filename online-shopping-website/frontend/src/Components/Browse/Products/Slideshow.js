import * as React from 'react';
import 'react-slideshow-image/dist/styles.css';
import { Slide } from "react-slideshow-image";
import {Link} from "react-router-dom";

const ProductShow = (props) => {
  return (
    <Link to={{
      pathname: `/${props.product.id}/${props.product.name}`,
      params: { props }
    }} className="RoutingLink">
      <div className="each-slide">
        <div>
          <div className="CenteringProductDetails">
            <div className="ProductDetailsShow">
              <img className="ProductImageShow" src={props.product.picture} alt={props.product.name}/>
              <div>
                <h3 className="ProductInfoSlide">{props.product.name}</h3>
                <p className="ProductInfoSlide">Brand: {props.product.brand.name}</p>
                <p className="ProductInfoSlide">Sold by: {props.product.seller.sellerName}</p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </Link>
  )
}

const iterateProducts = (data) => {
  return data.products.map(product => {
    return (
      <ProductShow key={product.id} product={product} />
    );
  });
}

export const Slideshow = (props) => {
  return (
    <div className="SlideshowPadding">
      <h1 className="PromotedItems">SPECIAL!</h1>
      <Slide easing="ease">
        {iterateProducts(props)}
      </Slide>
    </div>
  )
}