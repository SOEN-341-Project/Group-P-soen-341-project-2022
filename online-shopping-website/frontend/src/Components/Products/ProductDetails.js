import Grid from '@mui/material/Grid';
import Card from '@mui/material/Card';
import * as React from 'react';
import Button from '@mui/material/Button';
import AddShoppingCartIcon from '@mui/icons-material/AddShoppingCart';
import Stack from '@mui/material/Stack';
import AddIcon from '@mui/icons-material/Add';
import RemoveIcon from '@mui/icons-material/Remove';
import {useParams} from 'react-router-dom';
import Products from '../../TestValues.json';
import {Link} from "react-router-dom";


class ProductButtons extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            quantity: 1,
            show: true,
            max: 5,
            min: 0,
        };
    }

    IncrementItem = () => {
        if (this.state.quantity !== 10) {
            this.setState({
                quantity: this.state.quantity + 1
            });
        }
    }
    DecreaseItem = () => {
        if (this.state.quantity !== 1) {
            this.setState({quantity: this.state.quantity - 1});
        }
    }

    UpdateValue = (e) => {
        const inputValue = Number(e.target.value);

        if ((inputValue < 1) || (inputValue.isNaN())) {
            this.setState({quantity: 1});
        } else if (inputValue > 10) {
            this.setState({quantity: 10});
        } else {
            this.setState({quantity: inputValue});
        }
    }

    render() {
        return (
            <div className="ProductDetails-QuantityButtonsContainer">
                <h3>Quantity</h3>
                <Stack className="ProductDetails-QuantityButtonsStack" direction="row" spacing={1}>
                    <Button className="ProductDetails-QuantityButtons GreenButtonContained" variant="contained"
                            disabled={this.state.quantity === 1}
                            onClick={this.DecreaseItem}>
                        <RemoveIcon/>
                    </Button>
                    <input className="inputne" disabled={true} value={this.state.quantity} onChange={this.UpdateValue}/>
                    <Button className="ProductDetails-QuantityButtons GreenButtonContained" variant="contained"
                            disabled={this.state.quantity === 10}
                            onClick={this.IncrementItem}>
                        <AddIcon/>
                    </Button>
                </Stack>
                <h5 className="ProductDetails-ProductLimitText">Limit of 10 items per product in cart.</h5>
                <Link to="/my-shopping-cart" className="RoutingLink">
                    <Button className="ProductDetails-CartButton GreenButtonContained" variant="contained" endIcon={<AddShoppingCartIcon/>}>
                        Add to cart
                    </Button>
                </Link>
            </div>
        );
    }
}

export const ProductDetails = () => {
    //Resetting scrolling to top of the page
    window.scrollTo(0, 0);

    //Getting product name from URL
    const productId = useParams();

    //Getting product by id from URL
    const selectedProduct = Products.products.find(product => parseInt(productId.productId.match("[^/]*")) === product.id)

    return (
        <Grid container className="ProductDetails-Container">
            <Link to="/" className="RoutingLink">
                <Button className="GreenButtonContained" variant="contained">
                    Return to products
                </Button>
            </Link>

            <Grid item container xs={12} className="ProductDetails-ItemContainer">
                <Grid item xs={12} md={6}>
                    <h1>{selectedProduct.name}</h1>
                    <div className="ProductDetails-ImageConatiner">
                        <img className="ProductDetails-Image" src={selectedProduct.image} alt={selectedProduct.name}/>
                    </div>

                    <Grid item container>
                        <Grid item xs={12} md={6}>
                            <h3>Brand</h3>
                            <p>{selectedProduct.brand}</p>
                        </Grid>

                        <Grid item xs={12} md={6}>
                            <h3>Seller</h3>
                            <p>{selectedProduct.seller}</p>
                        </Grid>
                    </Grid>
                    <h3>Description</h3>
                    <p>{selectedProduct.description}</p>
                </Grid>
                <Grid item md={1}/>
                <Grid item xs={12} sm={12} md={5}>
                    <Card className="ProductDetails-SelectionPanel">
                        <h3>Price</h3>
                        <h4>{selectedProduct.price} Æ‰</h4>

                        <ProductButtons product={selectedProduct}/>
                    </Card>
                </Grid>
            </Grid>
        </Grid>
    );
}