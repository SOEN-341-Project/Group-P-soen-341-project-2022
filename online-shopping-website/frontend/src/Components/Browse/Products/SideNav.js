import * as React from 'react';
import Accordion from '@mui/material/Accordion';
import AccordionSummary from '@mui/material/AccordionSummary';
import AccordionDetails from '@mui/material/AccordionDetails';
import Typography from '@mui/material/Typography';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import Box from '@mui/material/Box';
import Slider from '@mui/material/Slider';
import Checkbox from '@mui/material/Checkbox';
import FormControlLabel from '@mui/material/FormControlLabel';
import { SearchBar } from './SearchBar';
import { Stack } from '@mui/material';

export const SideNav = (props) => {
    return (
        <Stack xs={12}>
            <SearchBar className="SearchBar" filterProducts={props.filterProducts} filters={props.filters} />
            <PriceFilter unfilteredProducts={props.unfilteredProducts} onSliderChange={props.onSliderChange} />
            <BrandDropdown brands={props.brands} onCheckboxChange={props.onCheckboxChange} />
            <SellerDropdown sellers={props.sellers} onCheckboxChange={props.onCheckboxChange} />
        </Stack>
    );
}

const PriceFilter = (props) => {
    
    function valuetext(value) {
        return `${value}Ɖ`;
    }
    
    // Gets lowest price of all products
    const getLowestPrice = (products) => {
        return Math.min.apply(Math, products.map((product) => { return product.price; }));
    }
    
    // Gets highest price of all products
    const getHighestPrice = (products) => {
        return Math.max.apply(Math, products.map((product) => { return product.price; }));
    }
    
    const lowestPrice = getLowestPrice(props.unfilteredProducts);
    const highestPrice = getHighestPrice(props.unfilteredProducts);
    // Labels to show under slider range
    const marks = [
        {
            value: lowestPrice,
            label: valuetext(lowestPrice),
        },
        {
            value: highestPrice,
            label: valuetext(highestPrice),
        },
    ];

    const [value, setValue] = React.useState([lowestPrice, highestPrice]);

    const handleChange = (event, newValue) => {
        setValue(newValue);
        props.onSliderChange(newValue);
    };

    return (
        <Box sx={{ display: 'flex', flexDirection: 'column', ml: 3, width: 140, paddingTop: '1rem', }}>
            <Typography>Filter by price:</Typography>
            <Typography>{value[0]}Ɖ - {value[1]}Ɖ</Typography>
            <Slider
                className='TextPink'
                getAriaLabel={() => 'Price range'}
                min={lowestPrice}
                max={highestPrice}
                value={value}
                onChange={handleChange}
                valueLabelDisplay="auto"
                getAriaValueText={valuetext}
                step={1}
                marks={marks}
                disableSwap
            />
        </Box>
    );
}

const BrandDropdown = (props) => {
    const handleChange = (event) => {
        props.onCheckboxChange('Brand', event.target.name, event.target.checked);
    };

    const iterateBrands = (props) => {
        return props.brands.map((brand,index) => {
            return (
                <Box key={index} sx={{ display: 'flex', flexDirection: 'column', ml: 3 }}>
                    <FormControlLabel
                        label={brand.name}
                        control={<Checkbox sx={{ '&.Mui-checked': { color: 'rgb(60, 121, 60)' } }}
                                           defaultChecked onChange={handleChange} name={brand.name} />}
                    />
                </Box>
            );
        });
    }

    return (
        <div className="accordion-width">
            <Accordion>
                <AccordionSummary
                    expandIcon={<ExpandMoreIcon />}
                    aria-controls="panel1a-content"
                    id="panel1a-header"
                >
                    <Typography>Brands</Typography>
                </AccordionSummary>
                <AccordionDetails>
                    <div>
                        {iterateBrands(props)}
                    </div>
                </AccordionDetails>
            </Accordion>
        </div>
    );
}

const SellerDropdown = (props) => {
    const handleChange = (event) => {
        props.onCheckboxChange('Seller', event.target.name, event.target.checked);
    };

    const iterateSellers = (props) => {
        return props.sellers.map((seller, index) => {
            return (
                <Box key={index} sx={{ display: 'flex', flexDirection: 'column', ml: 3 }}>
                    <FormControlLabel
                        label={seller.sellerName}
                        control={
                            <Checkbox
                                sx={{ '&.Mui-checked': { color: 'rgb(60, 121, 60)' } }}
                                defaultChecked
                                onChange={handleChange}
                                name={seller.sellerName}
                            />
                        }
                    />
                </Box>
            );
        });
    }
    return (
        <div className="accordion-width">
            <Accordion>
                <AccordionSummary
                    expandIcon={<ExpandMoreIcon />}
                    aria-controls="panel1a-content"
                    id="panel1a-header"
                >
                    <Typography>Sellers</Typography>
                </AccordionSummary>
                <AccordionDetails>
                    {iterateSellers(props)}
                </AccordionDetails>
            </Accordion>
        </div>
    );
}
