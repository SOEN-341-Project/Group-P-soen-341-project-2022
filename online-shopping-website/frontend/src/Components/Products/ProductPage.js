import { useState, useEffect } from 'react';
import Grid from '@mui/material/Grid';
import { ProductGrid } from './ProductGrid';
import { SideNav } from './SideNav';
import axios from 'axios';

// Encapsulates both SideNav and ProductGrid
export const ProductPage = () => {
  // Getting backend products
  const [products, setProducts] = useState(null);
  const [brands, setBrands] = useState(null);
  const [sellers, setSellers] = useState(null);
  const [loading, setLoading] = useState(true);

  const [unfilteredProducts, setUnfilteredProducts] = useState(null);

  useEffect(() => {
    const getProducts = axios.get(process.env.REACT_APP_DB_CONNECTION + '/api/items/all');
    const getBrands = axios.get(process.env.REACT_APP_DB_CONNECTION + '/api/brands/all');
    const getSellers = axios.get(process.env.REACT_APP_DB_CONNECTION + '/api/users/sellers');

    axios.all([getProducts, getBrands, getSellers]).then(
      axios.spread((resProducts, resBrands, resSellers) => {
        setProducts(resProducts.data);
        setUnfilteredProducts(resProducts.data);
        setBrands(resBrands.data);
        setSellers(resSellers.data);
        setLoading(false);
      })
    );
  }, []);

  // Filters
  let [filters] = useState({
    searchQuery: '',
    brands: [],
    sellers: [],
    lowestPrice: null,
    highestPrice: null,
  });

  const filterProducts = () => {
    // Use RegEx for search query
    /*
          ^ = Match beginning of product name
          i = Case insensitive
      */
    const query = new RegExp('^' + filters.searchQuery, 'i');

    // Filter products according to search query, are within price range, and don't belong to brands or sellers arrays
    setProducts(
      unfilteredProducts.filter(
        (product) =>
          (product.name.split(' ').some((word) => word.search(query) > -1) || product.name.search(query) > -1) &&
          (filters.lowestPrice ? product.price >= filters.lowestPrice : true) &&
          (filters.highestPrice ? product.price <= filters.highestPrice : true) &&
          !filters.brands.includes(product.brand.name) &&
          !filters.sellers.includes(product.seller.sellerName)
      )
    );
  };

  const onSliderChange = (priceRange) => {
    // Set price range from slider
    filters.lowestPrice = priceRange[0];
    filters.highestPrice = priceRange[1];

    // Update product state with filters
    filterProducts();
  };

  const onCheckboxChange = (filterType, filterName, isChecked) => {
    // Unchecked checkbox -> Add filter to filters state
    if (!isChecked) {
      if (filterType === 'Brand') {
        filters.brands.push(filterName);
      } else if (filterType === 'Seller') {
        filters.sellers.push(filterName);
      }
    }
    // Checked checkbox -> Remove filter from filters state
    else {
      if (filterType === 'Brand') {
        filters.brands = filters.brands.filter((f) => f !== filterName);
      } else if (filterType === 'Seller') {
        filters.sellers = filters.sellers.filter((f) => f !== filterName);
      }
    }

    // Update product state with filters
    filterProducts();
  };

  // Waiting for products during GET
  if (loading) {
    return <h1>Loading products...</h1>;
  }

  console.log("Brands: "+ brands + " , Sellers: " + sellers)
  return (
    <Grid container columnSpacing={4} rowSpacing={5}>
      <Grid item xs={12} md={3} lg={2}>
        <SideNav
          unfilteredProducts={unfilteredProducts}
          brands={brands}
          sellers={sellers}
          filterProducts={filterProducts}
          filters={filters}
          onCheckboxChange={onCheckboxChange}
          onSliderChange={onSliderChange}
        />
      </Grid>
      <Grid item xs={12} md={9} lg={10}>
        <ProductGrid products={products} />
      </Grid>
    </Grid>
  );
};
