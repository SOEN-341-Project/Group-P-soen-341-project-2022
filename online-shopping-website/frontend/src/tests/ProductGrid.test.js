/* eslint-disable no-undef */
import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import axios from "axios"
import { ProductGrid } from '../Components/Browse/Products/ProductGrid'
import { BrowserRouter } from "react-router-dom";

const testProducts = [
  {
    id: 1,
    name: 'thing 1',
    picture: 'https://www.irishtimes.com/polopoly_fs/1.3628113.1536842312!/image/image.jpg_gen/derivatives/ratio_1x1_w1200/image.jpg',
    brand: {name : 'Test product 1'},
    seller: {sellerName : 'Test seller 1'},
    price: 30
  },
  {
    id: 2,
    name: 'thing 2',
    picture: 'https://i.kym-cdn.com/entries/icons/original/000/033/421/cover2.jpg',
    brand: {name : 'Test product 2'},
    seller: {sellerName : 'Test seller 2'},
    price: 40
  }
];

describe('Product Grid Component', () => {
  // check if product grid has changed or not
  it('should render', () => {
    const component = render(<BrowserRouter><ProductGrid products={testProducts} /></BrowserRouter>);
    expect(component).toMatchSnapshot();
  });
  
  // check if iterate products works by seeing if all the products have their names in the product grid
  it('displays all product names', async () => {
    const products = (await axios.get('https://api.bobbleshop.me/api/items/all')).data;
    const component = render(<BrowserRouter><ProductGrid products={products} /></BrowserRouter>);
    const allProductNamesPresent = products.every(product => (component.getAllByText(product.name)).length > 0);
    expect(allProductNamesPresent).toBe(true);
  });
});