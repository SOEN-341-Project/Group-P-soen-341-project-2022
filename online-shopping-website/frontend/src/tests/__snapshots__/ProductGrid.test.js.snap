// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Grid Component should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 css-1ld3b9g-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-xl-4 css-70jto0-MuiGrid-root"
        >
          <a
            class="RoutingLink"
            href="/1/thing 1"
          >
            <button
              class="MuiButtonBase-root css-f8eae8-MuiButtonBase-root"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductContainer css-bhp9pd-MuiPaper-root-MuiCard-root"
              >
                <div
                  class="css-1shxafo-MuiStack-root"
                >
                  <div
                    class="css-nzye12-MuiStack-root"
                  >
                    <img
                      alt="thing 1"
                      class="ProductImage"
                      src="https://www.irishtimes.com/polopoly_fs/1.3628113.1536842312!/image/image.jpg_gen/derivatives/ratio_1x1_w1200/image.jpg"
                    />
                  </div>
                  <div>
                    <h3>
                      thing 1
                    </h3>
                    <p>
                      <strong>
                        Brand:
                      </strong>
                       
                      Test product 1
                    </p>
                    <p>
                      <strong>
                        Sold by:
                      </strong>
                       
                      Test seller 1
                    </p>
                  </div>
                  <h4
                    class="ProductPrice"
                  >
                    30
                     Ɖ
                  </h4>
                </div>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-xl-4 css-70jto0-MuiGrid-root"
        >
          <a
            class="RoutingLink"
            href="/2/thing 2"
          >
            <button
              class="MuiButtonBase-root css-f8eae8-MuiButtonBase-root"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductContainer css-bhp9pd-MuiPaper-root-MuiCard-root"
              >
                <div
                  class="css-1shxafo-MuiStack-root"
                >
                  <div
                    class="css-nzye12-MuiStack-root"
                  >
                    <img
                      alt="thing 2"
                      class="ProductImage"
                      src="https://i.kym-cdn.com/entries/icons/original/000/033/421/cover2.jpg"
                    />
                  </div>
                  <div>
                    <h3>
                      thing 2
                    </h3>
                    <p>
                      <strong>
                        Brand:
                      </strong>
                       
                      Test product 2
                    </p>
                    <p>
                      <strong>
                        Sold by:
                      </strong>
                       
                      Test seller 2
                    </p>
                  </div>
                  <h4
                    class="ProductPrice"
                  >
                    40
                     Ɖ
                  </h4>
                </div>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 css-1ld3b9g-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-xl-4 css-70jto0-MuiGrid-root"
      >
        <a
          class="RoutingLink"
          href="/1/thing 1"
        >
          <button
            class="MuiButtonBase-root css-f8eae8-MuiButtonBase-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductContainer css-bhp9pd-MuiPaper-root-MuiCard-root"
            >
              <div
                class="css-1shxafo-MuiStack-root"
              >
                <div
                  class="css-nzye12-MuiStack-root"
                >
                  <img
                    alt="thing 1"
                    class="ProductImage"
                    src="https://www.irishtimes.com/polopoly_fs/1.3628113.1536842312!/image/image.jpg_gen/derivatives/ratio_1x1_w1200/image.jpg"
                  />
                </div>
                <div>
                  <h3>
                    thing 1
                  </h3>
                  <p>
                    <strong>
                      Brand:
                    </strong>
                     
                    Test product 1
                  </p>
                  <p>
                    <strong>
                      Sold by:
                    </strong>
                     
                    Test seller 1
                  </p>
                </div>
                <h4
                  class="ProductPrice"
                >
                  30
                   Ɖ
                </h4>
              </div>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-xl-4 css-70jto0-MuiGrid-root"
      >
        <a
          class="RoutingLink"
          href="/2/thing 2"
        >
          <button
            class="MuiButtonBase-root css-f8eae8-MuiButtonBase-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ProductContainer css-bhp9pd-MuiPaper-root-MuiCard-root"
            >
              <div
                class="css-1shxafo-MuiStack-root"
              >
                <div
                  class="css-nzye12-MuiStack-root"
                >
                  <img
                    alt="thing 2"
                    class="ProductImage"
                    src="https://i.kym-cdn.com/entries/icons/original/000/033/421/cover2.jpg"
                  />
                </div>
                <div>
                  <h3>
                    thing 2
                  </h3>
                  <p>
                    <strong>
                      Brand:
                    </strong>
                     
                    Test product 2
                  </p>
                  <p>
                    <strong>
                      Sold by:
                    </strong>
                     
                    Test seller 2
                  </p>
                </div>
                <h4
                  class="ProductPrice"
                >
                  40
                   Ɖ
                </h4>
              </div>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
